internal/runtime/internal/controller/value_cache.go

```diff
@@ -1,88 +1,90 @@
+const argumentLabel = "argument"
+
+// valueCache caches exports and module arguments to expose as variables for Alloy expressions.
+// It also caches module exports to expose them to the parent loader.
+// The exports are stored directly in the scope which is used to evaluate Alloy expressions.
 type valueCache struct {
 	mut                sync.RWMutex
-	components         map[string]ComponentID // NodeID -> ComponentID
-	args               map[string]interface{} // NodeID -> component arguments value
-	exports            map[string]interface{} // NodeID -> component exports value
-	moduleArguments    map[string]any         // key -> module arguments value
-	moduleExports      map[string]any         // name -> value for the value of module exports
+	componentIds       map[string]ComponentID // NodeID -> ComponentID
+	moduleExports      map[string]any         // Export label -> Export value
+	moduleArguments    map[string]any         // Argument label -> Map with the key "value" that points to the Argument value
 	moduleChangedIndex int                    // Everytime a change occurs this is incremented
 	scope              *vm.Scope              // scope provides additional context for the nodes in the module
 }
 
 // newValueCache creates a new ValueCache.
 func newValueCache() *valueCache {
 	return &valueCache{
-		components:      make(map[string]ComponentID),
-		args:            make(map[string]interface{}),
-		exports:         make(map[string]interface{}),
-		moduleArguments: make(map[string]any),
+		componentIds:    make(map[string]ComponentID, 0),
 		moduleExports:   make(map[string]any),
+		moduleArguments: make(map[string]any),
+		scope:           vm.NewScope(make(map[string]any)),
 	}
 }
 
-func (vc *valueCache) SetScope(scope *vm.Scope) {
-	vc.mut.Lock()
-	defer vc.mut.Unlock()
-	vc.scope = scope
-}
-
-// CacheArguments will cache the provided arguments by the given id. args may
-// be nil to store an empty object.
-func (vc *valueCache) CacheArguments(id ComponentID, args component.Arguments) {
+// UpdateScopeVariables updates the Variables map of the scope with a deep copy of the provided map.
+func (vc *valueCache) UpdateScopeVariables(variables map[string]any) {
+	if variables == nil {
+		return
+	}
 	vc.mut.Lock()
 	defer vc.mut.Unlock()
-
-	nodeID := id.String()
-	vc.components[nodeID] = id
-
-	var argsVal interface{} = make(map[string]interface{})
-	if args != nil {
-		argsVal = args
-	}
-	vc.args[nodeID] = argsVal
+	vc.scope.Variables = deepCopyMap(variables)
 }
 
 // CacheExports will cache the provided exports using the given id. exports may
 // be nil to store an empty object.
-func (vc *valueCache) CacheExports(id ComponentID, exports component.Exports) {
+func (vc *valueCache) CacheExports(id ComponentID, exports component.Exports) error {
 	vc.mut.Lock()
 	defer vc.mut.Unlock()
 
-	nodeID := id.String()
-	vc.components[nodeID] = id
+	variables := vc.scope.Variables
+	// Build nested maps.
```

Comment:
Can you help me understand why do we choose to take on the complexity of the nested maps instead of using a flattened map with a concatenated component id parts, like "foo.bar.baz" instead of `{foo: { bar: baz}}`?

Reply:
because that's the structure that we need to give to the Alloy syntax to evaluate the expression. It's also the structure that we use for modules to propagate the context. The current implementation builds it on every evaluation whereas this new implementation stores the values directly in it